crates/bevy_reflect/bevy_reflect_derive/src/lib.rs

```diff
@@ -607,6 +605,198 @@ pub fn impl_reflect_value(input: TokenStream) -> TokenStream {
     )
 }
 
+/// Represents the information needed to implement a type as a Reflect Struct.
+/// # Example
+/// ```ignore
+/// impl_reflect_struct!(
+///    //                          attrs
+///    //        |----------------------------------------|
+///    #[reflect(PartialEq, Serialize, Deserialize, Default)]
+///    //            type_name       generics
+///    //     |-------------------||----------|
+///    struct ThingThatImReflecting<T1, T2, T3> {
+///        x: T1, // |
+///        y: T2, // |- fields
+///        z: T3  // |
+///    }
+/// );
+/// ```
+struct ReflectStructDef {
+    type_name: Ident,
+    generics: Generics,
+    attrs: ReflectAttrs,
+    fields: Fields,
+}
+
+impl Parse for ReflectStructDef {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let ast = input.parse::<DeriveInput>()?;
+
+        let type_name = ast.ident;
+        let generics = ast.generics;
+        let fields = match ast.data {
+            Data::Struct(data) => data.fields,
+            Data::Enum(data) => {
+                return Err(syn::Error::new_spanned(
+                    data.enum_token,
+                    "Enums are not currently supported for reflection",
+                ))
+            }
+            Data::Union(data) => {
+                return Err(syn::Error::new_spanned(
+                    data.union_token,
+                    "Unions are not supported for reflection",
+                ))
+            }
+        };
+
+        let mut attrs = ReflectAttrs::default();
+        for attribute in ast.attrs.iter().filter_map(|attr| attr.parse_meta().ok()) {
+            let meta_list = if let Meta::List(meta_list) = attribute {
+                meta_list
+            } else {
+                continue;
+            };
+
+            if let Some(ident) = meta_list.path.get_ident() {
+                if ident == REFLECT_ATTRIBUTE_NAME || ident == REFLECT_VALUE_ATTRIBUTE_NAME {
+                    attrs = ReflectAttrs::from_nested_metas(&meta_list.nested);
+                }
+            }
+        }
+
+        Ok(Self {
+            type_name,
+            generics,
+            attrs,
+            fields,
+        })
+    }
+}
+
+/// A replacement for `#[derive(Reflect)]` to be used with foreign types which
```

Comment:
Orphan rule is only an issue if you attempt to use this macro from outside bevy_reflect. In that case, you either need the type's owner to use #[derive(Reflect)] or you yourself need to use #[derive(Reflect)]. This macro is specifically for bevy to reflect standard foreign struct types, which so far is exclusively the glam types.

Though, this gives me an idea; wouldn't this macro work if applied to a newtype wrapper that has #[derive(Deref)], and you inform it of the layout of the inner, presumably foreign type? I'll have to try that. Could be a useful if niche trick

Reply:
Ha, works perfectly. That's some emergent design
```
#[derive(prelude::Deref, prelude::DerefMut, Default)]
struct BevyDMat4(self::math::DMat4);

use reflect::std_traits::ReflectDefault;
use prelude::Reflect;
use math::DVec4;

reflect::impl_reflect_struct!(
    #[reflect(Default)]
    struct BevyDMat4 {
        x_axis: DVec4,
        y_axis: DVec4,
        z_axis: DVec4,
        w_axis: DVec4
    }
);

#[test]
fn test() {
    let mut b = BevyDMat4(Default::default());
    b.x_axis.x = 12.0;

    let refl: &dyn Reflect = &b;

    use reflect::GetPath;
    let val = refl.path("x_axis.x").unwrap();
    assert_eq!(val.downcast_ref::<f64>().unwrap(), &12.0);
}
```

Reply:
Ooh interesting. This might be fairly useful actually since foreign types are a bit of an issue. It may even be worth adding an example for in `examples/reflection` (in a separate PR, of course).